Convert `px[ dDhHwWqQol]` to new shell

b1llow opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

convert `px[ dDhHwWqQol]` to new shell
- `pxd` -> `pxdw`
- `pxd 1` -> `pxd`
- `pxd 2` -> `pxdh`
- `pxd 4` -> `pxdw`
- `pxd 8` -> `pxdq`

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

b1llow committed:
convert `px[ dDhHwWqQol]` to new shell

- `pxd` -> `pxd 4`
- `pxd<N> [len]` -> `pxd <N> [len]`

b1llow committed:
fix

XVilka commented on librz/core/cmd/cmd_print.c:
```diff
@@ -6563,6 +6563,238 @@ RZ_IPI RzCmdStatus rz_print_utf32be_handler(RzCore *core, int argc, const char *
+	int len = argc > 2 ? (int)rz_num_math(core->num, argv[2]) : (int)core->blocksize;
+	return bool2status(print_dump(core, state, core->offset, n, len, RZ_CORE_PRINT_FORMAT_TYPE_INTEGER));
+}
+
```
Some of these functions could be converted into returning `char *` so that they would also be usable as an API, then you could also move them to [`librz/core/cprint.c`](https://github.com/rizinorg/rizin/blob/dev/librz/core/cprint.c).

b1llow replied:
But I feel that most of the existing APIs use `RzCons` to output, but it is not reasonable to change them to return `char*` one by one.

So I guess it might be reasonable to use something like this
https://github.com/rizinorg/cutter/blob/1dbad3d1edf3a2061de7fbe6e01af7d3da8494c8/src/dialogs/XrefsDialog.cpp#L134-L145

We can continue to use these APIs using `RzCons` instead of `rz_core_cmd*`

XVilka replied:
@imbillow, I don't think it's a good approach, more like a temporary solution. Also, Cutter isn't a sole user of the API, we wait to become a generally-usable library, even with various bindings; see https://github.com/rizinorg/rz-bindgen
I don't we have another choice than to return C structures wherever possible/make sense or strings where it doesn't.

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
expose APIs

- `rz_core_print_cmp`
- `rz_core_print_dump`
- `rz_core_print_hexdump_`
- `rz_core_print_hexdump_byline`

b1llow committed:
remove old

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
doxygen

XVilka commented:
@imbillow please rebase to fix the conflicts

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -429,6 +429,118 @@ commands:
+          - RZ_OUTPUT_MODE_JSON
+        args:
+          - name: len
+            type: RZ_CMD_ARG_TYPE_NUM
```
RZNUM

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -429,6 +429,118 @@ commands:
+            type: RZ_CMD_ARG_TYPE_NUM
+            optional: true
+          - name: len
+            type: RZ_CMD_ARG_TYPE_NUM
```
rznum

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -429,6 +429,118 @@ commands:
+          - RZ_OUTPUT_MODE_STANDARD
+          - RZ_OUTPUT_MODE_JSON
+        args:
+          - name: n
+            type: RZ_CMD_ARG_TYPE_NUM
+            optional: true
```
this is an option... another option is to mimic the behaviour we have right now, with subcommands like `pxd1`,`pxd2`, `pxd4`, etc. Like what is done for `wv?` for example.

ret2libc replied:
Actually... I wonder whether it would make sense to follow the same pattern as the regular `px` command... That is: `pxd`, `pxdh`, `pxdw`, `pxdq`. WDYT @XVilka @wargio ?

XVilka replied:
Yes, that would make more sense and be more consistent.

wargio replied:
yup

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -429,6 +429,118 @@ commands:
+          - RZ_OUTPUT_MODE_JSON
+        args:
+          - name: len
+            type: RZ_CMD_ARG_TYPE_NUM
```
please change all these `len` fields to RZNUM

ret2libc commented on librz/core/cprint.c:
```diff
@@ -30,4 +30,234 @@ RZ_API char *rz_core_print_string_c_cpp(RzCore *core) {
+ * \brief Print hexdump at \p addr, but maybe print hexdiff if (diff.from or diff.to), \see "el diff"
+ * \param len Dump bytes length
+ */
+RZ_API bool rz_core_print_hexdump_(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, int len) {
```
bad name ending with `_`

ret2libc commented on librz/core/cprint.c:
```diff
@@ -30,4 +30,234 @@ RZ_API char *rz_core_print_string_c_cpp(RzCore *core) {
+		return;
+	}
+	*len = -*len;
+	if (*len > core->blocksize_max) {
+		RZ_LOG_ERROR("this block size is too big (0x%" PFMT32x
+			     " < 0x%" PFMT32x ").",
+			*len,
+			core->blocksize_max);
+		*len = (int)core->blocksize_max;
+	}
```
shouldn't this be done before the initial `if`?

ret2libc commented on librz/core/cprint.c:
```diff
@@ -30,4 +30,234 @@ RZ_API char *rz_core_print_string_c_cpp(RzCore *core) {
+/**
+ * \brief Print hexdump diff between \p aa and \p ba with \p len
+ */
+RZ_API bool rz_core_print_cmp(RZ_NONNULL RzCore *core, ut64 aa, ut64 ba, ut64 len) {
```
maybe `rz_core_print_hexdump_diff`?

ret2libc commented on librz/core/cprint.c:
```diff
@@ -30,4 +30,234 @@ RZ_API char *rz_core_print_string_c_cpp(RzCore *core) {
+ * \param len Dump bytes length
+ * \param format Print format, such as RZ_CORE_PRINT_FORMAT_TYPE_HEXADECIMAL
+ */
+RZ_API bool rz_core_print_dump(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state,
```
in terms of output, isn't this supposed to be the same as `rz_core_print_hexdump`?

b1llow replied:
It's just a wrapper that supports {oct,hex,int}dump,

b1llow replied:
I actually don't understand why we have an API like the `rz_core_print_hexdump_or_hexdiff`, but just because it existed in the past, so I made it an API

XVilka replied:
You are absolutely right, if it's a printing function you can remove an RZ_API mark. That's what I usually do in my conversion PRs.

b1llow committed:
fix

b1llow commented on librz/core/cprint.c:
```diff
@@ -30,4 +30,237 @@ RZ_API char *rz_core_print_string_c_cpp(RzCore *core) {
+		char *vstr = ut64_to_hex(v, size * 2);
+		if (vstr) {
+			if (hex_offset) {
+				rz_print_section(core->print, addr + i);
```
It seems that there is no output to `sb` here, I have to find a way

b1llow committed:
fix

- test for `pxW` `pxH`
- `void rz_print_section` -> `const char *rz_print_section`

b1llow committed:
`pxd` -> `pxd[ hwq]`

b1llow committed:
`char*`

b1llow committed:
fix

wargio commented on librz/cons/cons.c:
```diff
@@ -1271,6 +1271,15 @@ RZ_API void rz_cons_strcat(const char *str) {
 	}
 }
 
+RZ_API bool rz_cons_print_own(RZ_NULLABLE char *str) {
```
doxygen

XVilka commented on librz/util/print.c:
```diff
@@ -192,15 +207,15 @@ RZ_API void rz_print_addr(RzPrint *p, ut64 addr) {
 			const char *fin = Color_RESET;
 			if (dec) {
-				printfmt("%s%s%s%s%c", pre, white, space, fin, ch);
+				sb_printf("%s%s%s%s%c", pre, white, space, fin, ch);
```
I suggest using `rz_strbuf_append[f]()` directly instead. `sb_printf()` just adds confusion.

XVilka commented on librz/util/print.c:
```diff
@@ -970,13 +993,13 @@ RZ_API void rz_print_hexdump(RzPrint *p, ut64 addr, const ut8 *buf, int len, int
 						? ' '
 						: buf[j];
-					rz_print_byte(p, "%c", j, ch);
+					sb_print_own(rz_print_byte_str(p, "%c", j, ch));
```
`sb_print_own()` used exactly twice. I think it's better to unroll the macro and use these three lines here directly.

XVilka commented on librz/util/print.c:
```diff
@@ -1112,61 +1136,63 @@ static ut8 *M(const ut8 *b, int len) {
 
 // TODO: add support for cursor
-RZ_API void rz_print_hexdiff(RzPrint *p, ut64 aa, const ut8 *_a, ut64 ba, const ut8 *_b, int len, int scndcol) {
+RZ_API char *rz_print_hexdiff_str(RzPrint *p, ut64 aa, const ut8 *_a, ut64 ba, const ut8 *_b, int len, int scndcol) {
```
Please add Doxygen.

b1llow committed:
unroll `sb_print*`

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
dox

b1llow committed:
notion

XVilka commented on librz/core/cprint.c:
```diff
@@ -156,4 +156,240 @@ RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len) {
+ * \param size Word size by bytes (1,2,4,8)
+ * \return Hexdump string
+ */
+RZ_API RZ_OWN char *rz_core_print_hexdump_byline_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state,
```
Since these functions aren't printing anymore, maybe we could give them slightly different name? Any suggestions? @ret2libc @wargio @thestr4ng3r @wingdeans

XVilka commented on librz/include/rz_util/rz_print.h:
```diff
@@ -137,25 +137,32 @@ RZ_API char *rz_print_hexpair(RzPrint *p, const char *str, int idx);
-RZ_API void rz_print_hexdump(RzPrint *p, ut64 addr, const ut8 *buf, int len, int base, int step, size_t zoomsz);
-RZ_API int rz_print_jsondump(RzPrint *p, const ut8 *buf, int len, int wordsize);
-RZ_API void rz_print_hexdiff(RzPrint *p, ut64 aa, const ut8 *a, ut64 ba, const ut8 *b, int len, int scndcol);
+RZ_API RZ_OWN char *rz_print_hexdump_str(RZ_NONNULL RzPrint *p, ut64 addr, RZ_NONNULL const ut8 *buf, int len, int base, int step, size_t zoomsz);
```
If you put `RZ_OWN` here, please also put it in the definition as well.

XVilka commented on librz/util/print.c:
```diff
@@ -366,27 +372,31 @@ RZ_API const char *rz_print_byte_color(RzPrint *p, int ch) {
-RZ_API void rz_print_byte(RzPrint *p, const char *fmt, int idx, ut8 ch) {
-	PrintfCallback printfmt = (PrintfCallback)(p ? p->cb_printf : libc_printf);
-#define print(x) printfmt("%s", x)
+RZ_API char *rz_print_byte_str(RzPrint *p, const char *fmt, int idx, ut8 ch) {
```
Doxygen.

XVilka commented on librz/util/print.c:
```diff
@@ -499,23 +509,42 @@ RZ_API void rz_print_set_screenbounds(RzPrint *p, ut64 addr) {
 }
 
-RZ_API void rz_print_section(RzPrint *p, ut64 at) {
+RZ_API const char *rz_print_section_str(RzPrint *p, ut64 at) {
```
Doxygen

XVilka commented on librz/util/print.c:
```diff
@@ -1499,7 +1546,8 @@ static bool ishexprefix(char *p) {
 }
 
-RZ_API char *rz_print_colorize_opcode(RzPrint *print, char *p, const char *reg, const char *num, bool partial_reset, ut64 func_addr) {
+RZ_API char *
```
Doxygen.

XVilka commented on librz/util/print.c:
```diff
@@ -1726,43 +1774,50 @@ RZ_API int rz_print_get_cursor(RzPrint *p) {
 	ut16 *buf16 = (ut16 *)buf;
 	ut32 *buf32 = (ut32 *)buf;
 	ut64 *buf64 = (ut64 *)buf;
 	// TODDO: support p==NULL too
```
I suggest removing this TODO

XVilka commented on librz/util/print.c:
```diff
@@ -1726,43 +1774,50 @@ RZ_API int rz_print_get_cursor(RzPrint *p) {
+ * \param wordsize Size of a word in bits
+ * \return Dump JSON string
+ */
+RZ_API char *rz_print_jsondump_str(RzPrint *p, const ut8 *buf, int len, int wordsize) {
```
Looks like `p` here is unused, maybe remove it?

b1llow replied:
![image](https://user-images.githubusercontent.com/24983369/180590592-5f48959a-2f9a-42ee-bb2d-31b9697d777c.png)
maybe not completely unused

XVilka commented on librz/cons/cons.c:
```diff
@@ -1271,6 +1271,18 @@ RZ_API void rz_cons_strcat(const char *str) {
+/**
+ * \brief Prints \p str to the RzCons, and ownership is transferred into.
+ */
+RZ_API bool rz_cons_print_own(RZ_NULLABLE char *str) {
```
When ownership is transferred, should we add `RZ_OWN` to the argument as well? `RZ_OWN RZ_NULLABLE char *str`? cc @thestr4ng3r

ret2libc replied:
yes, I think so

ret2libc replied:
However... I wonder why we need this API. We don't usually have these kind of APIs and it looks it is just used to make the thing easier to write.

I would rather just use rz_cons_print and have the free where the string is defined.

XVilka reviewed (changes requested):
Looks good now apart from few minor nitpicks.

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+        summary: show signed integer dump
+        subcommands:
+          - name: pxd
+            summary: show signed integer dump
```
```suggestion
            summary: show 1-byte integer dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+                type: RZ_CMD_ARG_TYPE_RZNUM
+                optional: true
+          - name: pxdh
+            summary: show signed integer dump
```
```suggestion
            summary: show 2-bytes integer dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+                type: RZ_CMD_ARG_TYPE_RZNUM
+                optional: true
+          - name: pxdw
+            summary: show signed integer dump
```
```suggestion
            summary: show 4-bytes integer dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+                type: RZ_CMD_ARG_TYPE_RZNUM
+                optional: true
+          - name: pxdq
+            summary: show signed integer dump
```
```suggestion
            summary: show 8-bytes integer dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+                type: RZ_CMD_ARG_TYPE_RZNUM
+                optional: true
+      - name: pxh
+        summary: show hexadecimal half-words dump (16bit)
```
```suggestion
        summary: show 2-bytes hexadecimal integers dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxH
+        summary: same as above, but one per line (16bit)
```
```suggestion
        summary: show 2-bytes hexadecimal integers dump, one per line
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxw
+        summary: show hexadecimal words dump (32bit)
```
```suggestion
        summary: show 4-bytes hexadecimal integers dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxW
+        summary: same as above, but one per line (32bit)
```
```suggestion
        summary: show 4-bytes hexadecimal integers dump, one per line
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxq
+        summary: show hexadecimal quad-words dump (64bit)
```
```suggestion
        summary: show 8-bytes hexadecimal integers dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxQ
+        summary: same as above, but one per line (64bit)
```
```suggestion
        summary: show 8-bytes hexadecimal integers dump, one per line
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxo
+        summary: show octal dump
```
```suggestion
        summary: show 1-byte octal integers dump
```

ret2libc commented on librz/core/cmd_descs/cmd_print.yaml:
```diff
@@ -477,6 +477,151 @@ commands:
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
+      - name: pxl
+        summary: display <N> lines (rows) of hexdump
```
```suggestion
        summary: display <N> lines of hexdump
```

ret2libc commented on librz/core/cmd/cmd_print.c:
```diff
@@ -6465,6 +6196,70 @@ RZ_IPI RzCmdStatus rz_print_utf32be_handler(RzCore *core, int argc, const char *
+	return bool2status(rz_core_print_hexdump_or_hexdiff(core, state, core->offset, core->print->cols * len));
+}
+
+RZ_IPI RzCmdStatus rz_print_hexdump_hex_comman_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state, ut8 n) {
```
typo
```suggestion
RZ_IPI RzCmdStatus rz_print_hexdump_hex_common_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state, ut8 n) {
```

ret2libc commented on librz/core/cmd/cmd_print.c:
```diff
@@ -6465,6 +6196,70 @@ RZ_IPI RzCmdStatus rz_print_utf32be_handler(RzCore *core, int argc, const char *
+	return rz_print_hexdump_hex_comman_handler(core, argc, argv, state, 8);
+}
+
+RZ_IPI RzCmdStatus rz_print_hexdump_hexl_comman_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state, ut8 n) {
```
typo
```suggestion
RZ_IPI RzCmdStatus rz_print_hexdump_hexl_common_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state, ut8 n) {
```

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -1053,12 +1053,29 @@ RZ_API bool rz_core_meta_string_add(RzCore *core, ut64 addr, ut64 size, RzStrEnc
 RZ_API RZ_OWN char *rz_core_assembly_of_hex(RzCore *core, ut8 *hex, int len);
 RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len);
 
+RZ_API RZ_OWN char *rz_core_print_hexdump_diff_str(RZ_NONNULL RzCore *core, ut64 aa, ut64 ba, ut64 len);
+RZ_API RZ_OWN char *rz_core_print_dump_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, ut8 n, int len, RzCorePrintFormatType format);
+RZ_API RZ_OWN char *rz_core_print_hexdump_or_hexdiff_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, int len);
+RZ_API RZ_OWN char *rz_core_print_hexdump_byline_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, int len, ut8 size);
```
These APIs are not like the others... When you accept a RzCmdStateOutput it is expected that when the outputmode == STANDARD you print into rz_cons, however these APIs just return a string. I'm ok to have separate APIs that return a string, but IMHO it should be a different kind of API.

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -1053,12 +1053,29 @@ RZ_API bool rz_core_meta_string_add(RzCore *core, ut64 addr, ut64 size, RzStrEnc
+RZ_API RZ_OWN char *rz_core_print_hexdump_or_hexdiff_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, int len);
+RZ_API RZ_OWN char *rz_core_print_hexdump_byline_str(RZ_NONNULL RzCore *core, RZ_NULLABLE RzCmdStateOutput *state, ut64 addr, int len, ut8 size);
+
+#define rz_core_print_hexdump_diff(...)       rz_cons_print_own(rz_core_print_hexdump_diff_str(__VA_ARGS__))
+#define rz_core_print_dump(...)               rz_cons_print_own(rz_core_print_dump_str(__VA_ARGS__))
+#define rz_core_print_hexdump_or_hexdiff(...) rz_cons_print_own(rz_core_print_hexdump_or_hexdiff_str(__VA_ARGS__))
+#define rz_core_print_hexdump_byline(...)     rz_cons_print_own(rz_core_print_hexdump_byline_str(__VA_ARGS__))
+
```
These are why the rz_cons_print_own are needed, but I'd rather do something differently rather than the rz_cons_print_own API which is a bit weird. But first, the API above needs to be implemented by respecting the RzCmdStateOutput

ret2libc commented on librz/include/rz_util/rz_print.h:
```diff
@@ -137,25 +137,32 @@ RZ_API char *rz_print_hexpair(RzPrint *p, const char *str, int idx);
 #define SEEFLAG    -2
 #define JSONOUTPUT -3
 
+#define rz_print_hexdump(...)  rz_cons_print_own(rz_print_hexdump_str(__VA_ARGS__))
+#define rz_print_jsondump(...) rz_cons_print_own(rz_print_jsondump_str(__VA_ARGS__))
+#define rz_print_hexdiff(...)  rz_cons_print_own(rz_print_hexdiff_str(__VA_ARGS__))
+#define rz_print_cursor(...)   rz_cons_print(rz_print_cursor_str(__VA_ARGS__))
+#define rz_print_addr(...)     rz_cons_print_own(rz_print_addr_str(__VA_ARGS__))
+#define rz_print_byte(...)     rz_cons_print_own(rz_print_byte_str(__VA_ARGS__))
+
```
I'd rather have these implemented as usual with:
```c
RZ_API RZ_OWN void rz_print_hexdump(RZ_NONNULL RzPrint *p, ut64 addr, RZ_NONNULL const ut8 *buf, int len, int base, int step, size_t zoomsz) { 
   char *s = rz_print_hexdump_str(.......);
   rz_cons_print(s);
   free(s);
   return XXX;
}
```
and so on.

ret2libc commented on test/db/json/json4:
```diff
@@ -15,10 +15,10 @@ psj
-pxd2j
-pxd4j
-pxd8j
+pxdj 1
+pxdj 2
+pxdj 4
+pxdj 8
```
```suggestion
pxdj
pxdhj
pxdwj
pxdqj
```

ret2libc reviewed (commented):
I added some comments here and there!

Thanks again for your work on this, I understand it is challenging to touch some of this code  (a lot of legacy stuff...). Seeing removals of cb_printf and a more logical code organization is great!

For the future, I would suggest you try to make smaller PRs or organize the commits of your PRs in a logical and organized way. Having smaller scopes to review makes it easier for us to review and will help to get the code merged faster. In general, the more you touch, the more comments you'll receive and the more adjustments you'll have to make, risking that it's going to take a lot of time.

For example, for this one an approach could have been to do first a PR that modifies RzPrint stuff and then once you have all the pieces done in RzPrint do another PR for the porting of commands to RzShell.

I hope that's helping and it makes sense!

b1llow committed:
Update librz/core/cmd_descs/cmd_print.yaml

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
Update librz/core/cmd_descs/cmd_print.yaml

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
Apply suggestions from code review

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
Apply suggestions from code review

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

b1llow committed:
unroll macro

b1llow committed:
revert style change

b1llow committed:
doxygen

b1llow committed:
fix

b1llow commented:
@ret2libc Yes, you are right.

XVilka reviewed (approved):
Please fix the JSON tests

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

ret2libc commented on librz/util/print.c:
```diff
@@ -221,28 +235,40 @@ RZ_API void rz_print_addr(RzPrint *p, ut64 addr) {
+	if (!string) {
+		return;
+	}
+	p->cb_printf("%s", string);
```
before it was something like:
```c
PrintfCallback printfmt = (PrintfCallback)(p ? p->cb_printf : libc_printf);
```
so this is changing a bit how the function worked.

ret2libc commented on librz/util/print.c:
```diff
@@ -499,23 +535,45 @@ RZ_API void rz_print_set_screenbounds(RzPrint *p, ut64 addr) {
+	if (!use_section) {
+		return "";
 	}
+	static char section[22];
```
this is not good. return a `RZ_OWN char *` if that's the only way

ret2libc commented on librz/util/print.c:
```diff
@@ -151,20 +151,34 @@ RZ_API bool rz_print_cursor_pointer(RzPrint *p, int cur, int len) {
+ * \param len The length of the byte.
+ * \param set (set ? '[' : ']') is used.
+ */
+RZ_API RZ_BORROW const char *rz_print_cursor_str(RZ_NULLABLE RzPrint *p, int cur, int len, int set) {
```
also this is not good. remove static stuff

ret2libc commented on librz/util/print.c:
```diff
@@ -1036,39 +1097,51 @@ RZ_API void rz_print_hexdump(RzPrint *p, ut64 addr, const ut8 *buf, int len, int
+	if (!str) {
+		return;
+	}
+	p->cb_printf("%s", str);
```
same as before. old code had `PrintfCallback printfmt = (PrintfCallback)printf;` and cb_printf was called only if set. This new code changes behaviour.

ret2libc commented on librz/util/print.c:
```diff
@@ -1111,62 +1184,87 @@ static ut8 *M(const ut8 *b, int len) {
+	if (!str) {
+		return;
+	}
+	p->cb_printf("%s", str);
```
same

ret2libc commented on librz/include/rz_util/rz_print.h:
```diff
@@ -137,20 +137,23 @@ RZ_API char *rz_print_hexpair(RzPrint *p, const char *str, int idx);
-RZ_API void rz_print_hexdump(RzPrint *p, ut64 addr, const ut8 *buf, int len, int base, int step, size_t zoomsz);
-RZ_API int rz_print_jsondump(RzPrint *p, const ut8 *buf, int len, int wordsize);
-RZ_API void rz_print_hexdiff(RzPrint *p, ut64 aa, const ut8 *a, ut64 ba, const ut8 *b, int len, int scndcol);
+RZ_API RZ_OWN char *rz_print_hexdump_str(RZ_NONNULL RzPrint *p, ut64 addr, RZ_NONNULL const ut8 *buf, int len, int base, int step, size_t zoomsz);
```
I guess you are converting these functions into `_str` ones because you plan to re-use them in Cutter? How do you plan to do so?

I'm asking because I feel like these functions are very rizin-specific and I'm not sure how much of them is going to be useful outside of it (e.g. in cutter).

XVilka replied:
@ret2libc this is how `px` commands are used in Cutter: https://github.com/rizinorg/cutter/blob/dev/src/core/Cutter.cpp#L4194

ret2libc commented on librz/core/cprint.c:
```diff
@@ -156,4 +156,265 @@ RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len) {
+/**
+ * \brief Prints \p str to the RzCons, and ownership is transferred into.
+ */
+static inline bool rz_cons_print_own(char *str) {
```
no `rz_` prefix for static functions.

XVilka commented on librz/include/rz_util/rz_print.h:
```diff
@@ -137,20 +137,23 @@ RZ_API char *rz_print_hexpair(RzPrint *p, const char *str, int idx);
-RZ_API int rz_print_jsondump(RzPrint *p, const ut8 *buf, int len, int wordsize);
-RZ_API void rz_print_hexdiff(RzPrint *p, ut64 aa, const ut8 *a, ut64 ba, const ut8 *b, int len, int scndcol);
+RZ_API RZ_OWN char *rz_print_hexdump_str(RZ_NONNULL RzPrint *p, ut64 addr, RZ_NONNULL const ut8 *buf, int len, int base, int step, size_t zoomsz);
+RZ_API void rz_print_hexdump(RZ_NONNULL RzPrint *p, ut64 addr, RZ_NONNULL const ut8 *buf, int len, int base, int step, size_t zoomsz);
```
I suggest removing the `rz_print_hexdump()` and `rz_print_addr()`, and other, similar function from the `RZ_API` and substitute them with what you call `rz_print_hexdump_str()`, since as an API printing something to the console directly is useless.

XVilka commented on librz/util/print.c:
```diff
@@ -1726,43 +1824,31 @@ RZ_API int rz_print_get_cursor(RzPrint *p) {
+ * \param wordsize Size of a word in bits
+ * \return Dump JSON string
+ */
+RZ_API RZ_OWN char *rz_print_jsondump_str(RZ_NONNULL RzPrint *p, RZ_NONNULL const ut8 *buf, int len, int wordsize) {
```
This function could be probably removed from `librz/util/` and put directly in the `px` handler code, which should use `pj` allocated by the `RzCmdStateOutput`

XVilka commented on librz/util/print.c:
```diff
@@ -1111,62 +1184,87 @@ static ut8 *M(const ut8 *b, int len) {
+}
+
+/// \see rz_print_hexdiff_str
+RZ_API void rz_print_hexdiff(RZ_NONNULL RzPrint *p, ut64 aa, RZ_NONNULL const ut8 *_a, ut64 ba, RZ_NONNULL const ut8 *_b, int len, int scndcol) {
```
I suggest removing this function from `librz/util/` and use `rz_print_hexdiff_str()` wherever appropriate directly.

XVilka commented on librz/util/print.c:
```diff
@@ -1036,39 +1097,51 @@ RZ_API void rz_print_hexdump(RzPrint *p, ut64 addr, const ut8 *buf, int len, int
+}
+
+/// \see rz_print_hexdump_str
+RZ_API void rz_print_hexdump(RZ_NONNULL RzPrint *p, ut64 addr, RZ_NONNULL const ut8 *buf, int len, int base, int step, size_t zoomsz) {
```
I suggest removing this function from `librz/util`

XVilka commented on librz/util/print.c:
```diff
@@ -221,28 +235,40 @@ RZ_API void rz_print_addr(RzPrint *p, ut64 addr) {
+}
+
+/// \see rz_print_addr_str
+RZ_API void rz_print_addr(RZ_NONNULL RzPrint *p, ut64 addr) {
```
I suggest removing this function from `librz/util/`

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -1053,12 +1053,24 @@ RZ_API bool rz_core_meta_string_add(RzCore *core, ut64 addr, ut64 size, RzStrEnc
 RZ_API RZ_OWN char *rz_core_assembly_of_hex(RzCore *core, ut8 *hex, int len);
 RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len);
 
+RZ_API RZ_OWN char *rz_core_print_hexdump_diff_str(RZ_NONNULL RzCore *core, ut64 aa, ut64 ba, ut64 len);
+RZ_API RZ_OWN char *rz_core_print_dump_str(RZ_NONNULL RzCore *core, RzOutputMode mode, ut64 addr, ut8 n, int len, RzCorePrintFormatType format);
+RZ_API RZ_OWN char *rz_core_print_hexdump_or_hexdiff_str(RZ_NONNULL RzCore *core, RzOutputMode mode, ut64 addr, int len);
+RZ_API RZ_OWN char *rz_core_print_hexdump_byline_str(RZ_NONNULL RzCore *core, bool hex_offset, ut64 addr, int len, ut8 size);
```
these functions should be more like:
```
RZ_API bool rz_core_binfiles_print(RzCore *core, RzCmdStateOutput *state);
```

They should return a `bool` and accept a `RzCmdStateOutput` and directly print.

XVilka reviewed (changes requested):
The review got quite long, I wonder if it makes sense to send a new PR instead and close this one. @ret2libc what do you think?

XVilka commented:
In general, we want to eliminate all printing from the `librz/util/`, and print only from `librz/core/`. I am with @ret2libc on `RZ_API bool rz_core_binfiles_print(RzCore *core, RzCmdStateOutput *state);` but it probably should not be even an API at all - just `RZ_IPI`

ret2libc commented:
If this gets closed IMHO the new one should be smaller and more focused on one thing so that it's easier to review and merge

b1llow commented:
OK, let me cut it into smaller pieces
